libkbfs/file_data.go

```diff
@@ -97,6 +122,341 @@ func (fd *fileData) getFileBlockAtOffset(ctx context.Context,
+			for _, p := range pfr {
+				newPath := append([]parentBlockAndChildIndex{{
+					pblock:     pblock,
+					childIndex: childIndex,
+				}}, p...)
+				r.pathsFromRoot = append(r.pathsFromRoot, newPath)
+			}
+			r.blocks = make(map[BlockPointer]*FileBlock)
+			r.blocks[ptr] = block
+			for ptr, block := range blocks {
+				r.blocks[ptr] = block
+			}
+			r.nextBlockOff = nextBlockOff
+			respCh <- r
+			return nil
+		})
+	}
+
+	// If we are ok with just getting the prefix, don't treat a
+	// deadline exceeded error as fatal.
+	if err := eg.Wait(); err != nil && (!prefixOk ||
+		err != context.DeadlineExceeded) {
+		return nil, nil, 0, err
+	}
+
+	blocks = make(map[BlockPointer]*FileBlock)
+outer:
+	for i, respCh := range respChans {
+		select {
+		case r := <-respCh:
+			pathsFromRoot = append(pathsFromRoot, r.pathsFromRoot...)
+			for ptr, block := range r.blocks {
+				blocks[ptr] = block
+			}
+			if i == len(respChans)-1 {
+				// Figure out the offset of the next data block after
+				// the last in-range one.  It's the minimum of the
+				// next one at this level and the next one at the
+				// child's level, not counting -1 (which means we are
+				// returning the the last block at the given level).
+				if (r.nextBlockOff != -1 && r.nextBlockOff < nextBlockOff) ||
+					nextBlockOff == -1 {
+					nextBlockOff = r.nextBlockOff
+				}
+			}
+		default:
+			// There should always be a response ready in every
+			// channel, unless prefixOk is true.
+			if prefixOk {
+				break outer
+			} else {
+				panic("No response ready when !prefixOk")
+			}
+		}
+	}
+	return pathsFromRoot, blocks, nextBlockOff, nil
+}
+
+// getByteSlicesInOffsetRange returns an ordered, continuous slice of
+// byte ranges for the data described by the half-inclusive offset
+// range `[startOff, endOff)`.  If `endOff` == -1, it returns data to
+// the end of the file.  The caller is responsible for concatenating
+// the data into a single buffer if desired. If `prefixOk` is true,
+// the function will ignore context deadline errors and return
+// whatever prefix of the data it could fetch within the deadine.
+func (fd *fileData) getByteSlicesInOffsetRange(ctx context.Context,
+	startOff, endOff int64, prefixOk bool) ([][]byte, error) {
+	if (endOff != -1 && endOff <= startOff) || startOff < 0 {
+		return nil, nil
+	}
+
+	topBlock, _, err := fd.getter(ctx, fd.kmd, fd.file.tailPointer(),
+		fd.file, blockRead)
+	if err != nil {
+		return nil, err
+	}
+
+	// Find all the indirect pointers to leaf blocks in the offset range.
+	var iptrs []IndirectFilePtr
+	firstBlockOff := int64(-1)
+	endBlockOff := int64(-1)
+	nextBlockOff := int64(-1)
+	var blockMap map[BlockPointer]*FileBlock
+	if topBlock.IsInd {
+		var pfr [][]parentBlockAndChildIndex
+		pfr, blockMap, nextBlockOff, err = fd.getLeafBlocksForOffsetRange(
+			ctx, topBlock, startOff, endOff, prefixOk)
+		if err != nil {
+			return nil, err
+		}
+
+		for i, p := range pfr {
+			if len(p) == 0 {
+				continue
```

Comment:
seems like an error case, since an empty `p` would imply that `topBlock` is a direct block